core/src/local/persistent_state.rs

```diff
@@ -0,0 +1,891 @@
+            self.db.as_ref().unwrap().write(b).unwrap();
+        }
+
+        self.indices.push(PersistentIndex { columns: cols, seq });
+        self.persist_meta();
+    }
+
+    fn keys(&self) -> Vec<Vec<usize>> {
+        self.indices
+            .iter()
+            .map(|index| index.columns.clone())
+            .collect()
+    }
+
+    fn cloned_records(&self) -> Vec<Vec<DataType>> {
+        self.all_rows()
+            .map(|(_, ref value)| bincode::deserialize(&value).unwrap())
+            .collect()
+    }
+
+    fn rows(&self) -> usize {
+        self.all_rows().count()
+    }
+
+    fn clear(&mut self) {
+        // Would potentially be faster to just drop self.db and call DB::Destroy:
+        let db = self.db.as_ref().unwrap();
+        for (key, _) in db.full_iterator(rocksdb::IteratorMode::Start) {
+            db.delete(&key).unwrap();
+        }
+    }
+
+    fn is_useful(&self) -> bool {
+        self.indices.len() > 0
+    }
+
+    fn is_partial(&self) -> bool {
+        false
+    }
+
+    fn mark_filled(&mut self, _: Vec<DataType>, _: &Tag) {
+        unreachable!("PersistentState can't be partial")
+    }
+
+    fn mark_hole(&mut self, _: &[DataType], _: &Tag) {
+        unreachable!("PersistentState can't be partial")
+    }
+
+    fn evict_random_keys(&mut self, _: usize) -> (&[usize], Vec<Vec<DataType>>, u64) {
+        unreachable!("can't evict keys from PersistentState")
+    }
+
+    fn evict_keys(&mut self, _: &Tag, _: &[Vec<DataType>]) -> (&[usize], u64) {
+        unreachable!("can't evict keys from PersistentState")
+    }
+}
+
+impl PersistentState {
+    pub fn new(
+        name: String,
+        primary_key: Option<&[usize]>,
+        params: &PersistenceParameters,
+    ) -> Self {
+        let mut opts = rocksdb::Options::default();
+        opts.set_compression_type(rocksdb::DBCompressionType::Lz4);
+        opts.create_if_missing(true);
+        opts.create_missing_column_families(true);
+        let mut block_opts = rocksdb::BlockBasedOptions::default();
+        block_opts.set_bloom_filter(10, true);
+        opts.set_block_based_table_factory(&block_opts);
+
+        if let Some(ref path) = params.log_dir {
+            // Append the db name to the WAL path to ensure
+            // that we create a directory for each base shard:
+            opts.set_wal_dir(path.join(&name));
+        }
+
+        // Create prefixes by Self::transform_fn on all new inserted keys:
+        let transform = SliceTransform::create("key", Self::transform_fn, Some(Self::in_domain_fn));
+        opts.set_prefix_extractor(transform);
+
+        // Assigns the number of threads for RocksDB's low priority background pool:
+        opts.increase_parallelism(params.persistence_threads);
+
+        // Use a hash linked list since we're doing prefix seeks.
+        opts.set_allow_concurrent_memtable_write(false);
+        opts.set_memtable_factory(rocksdb::MemtableFactory::HashLinkList {
+            bucket_count: 1_000_000,
```

Comment:
Is this never resized, even if it becomes full?

Reply:
Yeah, the concern I specifically have is that hash maps stray further and further from `O(1)` the larger the buckets grow. If they never resize, then we could end up in pretty sad territory.